File: R/fusion.R
# Merge observations from multiple sources with conflict resolution.
# R twin of python/dssatcalibrator/fusion.py.

#' Extended observation schema (adds provenance columns to SCHEMA).
#' @export
SCHEMA_EXTENDED <- c("exp_id", "treatment", "variable", "kind", "date", "value",
                     "sigma", "weight", "source", "quality_flag", "spatial_res_m")

# one extended-schema row as a data.frame (used by the source adapters)
.ext_row <- function(exp_id, treatment, variable, kind, date, value, sigma,
                     weight, source, quality_flag, spatial_res_m) {
  data.frame(exp_id = as.character(exp_id), treatment = as.integer(treatment),
             variable = as.character(variable), kind = as.character(kind),
             date = as.Date(date), value = as.numeric(value), sigma = as.numeric(sigma),
             weight = as.numeric(weight), source = as.character(source),
             quality_flag = as.integer(quality_flag),
             spatial_res_m = as.numeric(spatial_res_m), stringsAsFactors = FALSE)
}

.empty_extended <- function() {
  data.frame(exp_id = character(0), treatment = integer(0), variable = character(0),
             kind = character(0), date = as.Date(character(0)), value = numeric(0),
             sigma = numeric(0), weight = numeric(0), source = character(0),
             quality_flag = integer(0), spatial_res_m = numeric(0),
             stringsAsFactors = FALSE)
}

#' Construct an ObservationFuser over a list of sources.
#' Mirrors fusion.py:ObservationFuser.__init__.
#' @export
observation_fuser <- function(sources, cfg) {
  named <- list()
  for (s in sources) named[[s$name]] <- s
  structure(list(sources = named, cfg = cfg), class = "observation_fuser")
}

#' Gather from all active sources, apply QC, merge. Mirrors ObservationFuser.collect.
#' @export
fuser_collect <- function(fuser, experiment, date_range) {
  frames <- list()
  for (name in names(fuser$sources)) {
    src <- fuser$sources[[name]]
    res <- tryCatch({
      df <- src_fetch(src, experiment, date_range)
      if (nrow(df) == 0) NULL else {
        df <- src_quality_filter(src, df)
        df$source <- name
        if (!("sigma" %in% names(df)) || any(is.na(df$sigma))) {
          df$sigma <- mapply(function(v, x) src_error_model(src, v, x, list()),
                             df$variable, df$value)
        }
        for (col in SCHEMA_EXTENDED) if (!(col %in% names(df))) df[[col]] <- NA
        df[SCHEMA_EXTENDED]
      }
    }, error = function(e) {
      warning(sprintf("Source %s failed to fetch for %s: %s", name, experiment,
                      conditionMessage(e)))
      NULL
    })
    if (!is.null(res)) frames[[length(frames) + 1L]] <- res
  }
  if (length(frames) == 0) return(.empty_extended())
  merged <- do.call(rbind, frames)
  fuser_resolve_conflicts(fuser, merged)
}

#' Handle overlapping observations from different sources.
#' Strategies: keep_all | inverse_variance | priority. Mirrors resolve_conflicts.
#' @export
fuser_resolve_conflicts <- function(fuser, df) {
  if (nrow(df) == 0) return(df)
  strategy <- .cfg_get(.cfg_get(fuser$cfg, "fusion", list()),
                       "conflict_resolution", "keep_all")
  if (strategy == "inverse_variance") return(.inverse_variance_merge(df))
  if (strategy == "priority") return(.priority_merge(fuser, df))
  df  # keep_all (and unknown) pass through
}

# Inverse-variance weighted merge of coincident observations.
.inverse_variance_merge <- function(df) {
  df$date_key <- as.Date(df$date)
  key <- paste(df$exp_id, df$treatment, df$variable, as.character(df$date_key), sep = "\r")
  out <- list()
  for (k in unique(key)) {
    g <- df[key == k, , drop = FALSE]
    if (nrow(g) == 1) {
      out[[length(out) + 1L]] <- g[, setdiff(names(g), "date_key"), drop = FALSE]
      next
    }
    gc <- g[!is.na(g$sigma), , drop = FALSE]
    if (nrow(gc) == 0) {
      out[[length(out) + 1L]] <- g[1, setdiff(names(g), "date_key"), drop = FALSE]
      next
    }
    sigmas <- gc$sigma; sigmas[sigmas == 0] <- 1e-6
    w <- 1.0 / (sigmas^2)
    val <- sum(gc$value * w) / sum(w)
    sig <- 1.0 / sqrt(sum(w))
    row <- gc[1, , drop = FALSE]
    row$value <- val; row$sigma <- sig
    row$source <- paste(sort(unique(gc$source)), collapse = "+")
    row$quality_flag <- as.integer(max(gc$quality_flag))
    row$spatial_res_m <- mean(gc$spatial_res_m)
    out[[length(out) + 1L]] <- row[, setdiff(names(row), "date_key"), drop = FALSE]
  }
  do.call(rbind, out)
}

# Keep only the highest-priority source for coincident measurements.
.priority_merge <- function(fuser, df) {
  priority_list <- .cfg_get(.cfg_get(fuser$cfg, "fusion", list()), "source_priority", list())
  if (length(priority_list) == 0) return(df)
  priority_list <- unlist(priority_list, use.names = FALSE)
  p_map <- setNames(seq_along(priority_list) - 1L, priority_list)
  df$date_key <- as.Date(df$date)
  df$priority_rank <- vapply(df$source, function(x) {
    if (!is.na(p_map[x])) as.integer(p_map[[x]]) else 9999L
  }, integer(1))
  df <- df[order(df$priority_rank), , drop = FALSE]
  dedup_key <- paste(df$exp_id, df$treatment, df$variable, as.character(df$date_key), sep = "\r")
  keep <- !duplicated(dedup_key)
  df <- df[keep, , drop = FALSE]
  df[, setdiff(names(df), c("date_key", "priority_rank")), drop = FALSE]
}

File: python/dssatcalibrator/fusion.py
import logging
from datetime import date
import pandas as pd
import numpy as np
from .sources import ObservationSource

logger = logging.getLogger(__name__)

SCHEMA_EXTENDED = [
    "exp_id", "treatment", "variable", "kind", "date", "value", 
    "sigma", "weight", "source", "quality_flag", "spatial_res_m"
]

class ObservationFuser:
    """Merges observations from multiple sources with conflict resolution."""
    
    def __init__(self, sources: list[ObservationSource], cfg: dict):
        self.sources = {s.name: s for s in sources}
        self.cfg = cfg
    
    def collect(self, experiment: str, date_range: tuple[date, date], **kwargs) -> pd.DataFrame:
        """Gather from all active sources, apply QC, merge."""
        frames = []
        for name, src in self.sources.items():
            try:
                df = src.fetch(experiment, date_range, **kwargs)
                if df.empty:
                    continue
                df = src.quality_filter(df)
                df["source"] = name
                df["source_type"] = src.source_type
                
                # Apply source-specific sigma if not already set or is NaN
                if "sigma" not in df.columns or df["sigma"].isna().any():
                    df["sigma"] = df.apply(
                        lambda r: src.error_model(r["variable"], r["value"], r.get("metadata", {})),
                        axis=1
                    )
                
                # Ensure all SCHEMA_EXTENDED columns exist
                for col in SCHEMA_EXTENDED:
                    if col not in df.columns:
                        df[col] = np.nan
                
                frames.append(df[SCHEMA_EXTENDED])
            except Exception as e:
                logger.warning(f"Source {name} failed to fetch for {experiment}: {e}")
        
        if not frames:
            return pd.DataFrame(columns=SCHEMA_EXTENDED)
        
        merged = pd.concat(frames, ignore_index=True)
        return self.resolve_conflicts(merged)
    
    def resolve_conflicts(self, df: pd.DataFrame) -> pd.DataFrame:
        """Handle overlapping observations from different sources.
        
        Strategy:
        - "keep_all": Keep all observations
        - "inverse_variance": Weight by 1/sigma² (Bayesian optimal)
        - "priority": Use highest-priority source, discard others
        """
        if df.empty:
            return df
            
        strategy = self.cfg.get("fusion", {}).get("conflict_resolution", "keep_all")
        if strategy == "keep_all":
            return df
        elif strategy == "inverse_variance":
            return self._inverse_variance_merge(df)
        elif strategy == "priority":
            return self._priority_merge(df)
        return df
    
    def _inverse_variance_merge(self, df: pd.DataFrame) -> pd.DataFrame:
        """Merge coincident observations using inverse-variance weighting."""
        df_copy = df.copy()
        df_copy["date_key"] = df_copy["date"].dt.date
        
        groups = df_copy.groupby(["exp_id", "treatment", "variable", "date_key"])
        merged = []
        for key, g in groups:
            if len(g) == 1:
                merged.append(g.iloc[0].drop("date_key"))
            else:
                g_clean = g.dropna(subset=["sigma"])
                if g_clean.empty:
                    merged.append(g.iloc[0].drop("date_key"))
                    continue
                
                sigmas = np.where(g_clean["sigma"] == 0, 1e-6, g_clean["sigma"])
                w = 1.0 / (sigmas ** 2)
                val = (g_clean["value"] * w).sum() / w.sum()
                sig = 1.0 / np.sqrt(w.sum())
                
                row = g_clean.iloc[0].copy()
                row["value"] = val
                row["sigma"] = sig
                row["source"] = "+".join(sorted(g_clean["source"].unique()))
                row["quality_flag"] = int(g_clean["quality_flag"].max())
                row["spatial_res_m"] = g_clean["spatial_res_m"].mean()
                merged.append(row.drop("date_key"))
                
        return pd.DataFrame(merged)
        
    def _priority_merge(self, df: pd.DataFrame) -> pd.DataFrame:
        """Keep only the highest priority source for coincident measurements."""
        priority_list = self.cfg.get("fusion", {}).get("source_priority", [])
        if not priority_list:
            return df
            
        p_map = {name: i for i, name in enumerate(priority_list)}
        
        df_copy = df.copy()
        df_copy["date_key"] = df_copy["date"].dt.date
        df_copy["priority_rank"] = df_copy["source"].map(lambda x: p_map.get(x, 9999))
        
        df_copy = df_copy.sort_values("priority_rank")
        
        dedup = df_copy.drop_duplicates(subset=["exp_id", "treatment", "variable", "date_key"], keep="first")
        return dedup.drop(columns=["date_key", "priority_rank"])
